include/circt/Dialect/HW/HWOpInterfaces.td

```diff
@@ -15,6 +15,22 @@
 
 include "mlir/IR/SymbolInterfaces.td"
 include "mlir/IR/OpBase.td"
+def ModuleLike : OpInterface<"ModuleLike", []> {
+  let cppNamespace = "circt::hw";
+  let description = "Provide common module information.";
+
+  let methods = [
+    InterfaceMethod<"Get the module name",
+    "::llvm::StringRef", "getModuleLikeName", (ins),
+    /*methodBody=*/[{}],
+    /*defaultImplementation=*/[{ return $_op.getName(); }]>,
+
+    InterfaceMethod<"Get the module name",
+    "::mlir::StringAttr", "getModuleLikeNameAttr", (ins),
+    /*methodBody=*/[{}],
+    /*defaultImplementation=*/[{ return $_op.getNameAttr(); }]>,
+  ];
+}
 
 def HWModuleLike : OpInterface<"HWModuleLike", [Symbol]> {
```

Comment:
Should this not extend ModuleLike, which in turn extends Symbol? Would tblgen like that more?

Reply:
No, it hates it.  It winds up making duplicate copies of Symbol, which doesn't compile.  This is also why the names are slightly offset.

Reply:
Well, it doesn't seem to mind `HWModuleLike` extending `ModuleLike` and `Symbol`. Though actually the only thing that really needs (as defined as not barfing) to inherit `Symbol` is `FModuleLike`. Moving `Symbol` there would reduce the number of things that other stuff is not explicitly implementing...

```diff
diff --git a/include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.td b/include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.td
index 07b06a815..52fe47a82 100644
--- a/include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.td
+++ b/include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.td
@@ -16,7 +16,7 @@
 include "mlir/IR/OpBase.td"
 include "circt/Dialect/HW/HWOpInterfaces.td"

-def FModuleLike : OpInterface<"FModuleLike", [ModuleLike, HWModuleLike]> {
+def FModuleLike : OpInterface<"FModuleLike", [HWModuleLike, Symbol]> {
   let cppNamespace = "circt::firrtl";
   let description = "Provide common module information.";
   let methods = [
diff --git a/include/circt/Dialect/HW/HWOpInterfaces.td b/include/circt/Dialect/HW/HWOpInterfaces.td
index 4da854411..c41f13339 100644
--- a/include/circt/Dialect/HW/HWOpInterfaces.td
+++ b/include/circt/Dialect/HW/HWOpInterfaces.td
@@ -32,7 +32,7 @@ def ModuleLike : OpInterface<"ModuleLike", []> {
   ];
 }

-def HWModuleLike : OpInterface<"HWModuleLike", [Symbol]> {
+def HWModuleLike : OpInterface<"HWModuleLike", [ModuleLike]> {
   let cppNamespace = "circt::hw";
   let description = "Provide common module information.";
```

Reply:
Just inheriting from `ModuleLike` would reduce the number of aliases for module name by 2.

Reply:
https://github.com/llvm/circt/compare/dev/dathscsi/NarrowIG...dev/teqdruid/NarrowIG?expand=1